### DISCUSSION TITLE
Unpacking multiple `TypeVarTuples`

### DISCUSSION THREAD

**Discussion by Zomatree** [python/typing#1439] [General]
[PEP 646 says that you cannot unpack multiple `TypeVarTuples` at once](https://peps.python.org/pep-0646/#multiple-unpackings-in-a-tuple-not-allowed), but i cant think of any reason for this limitation, there shouldn't be any ambiguity in it like with having multiple variadic generics which what the reason links to.

A simple use-case for this is combining tuples together:
```python
Ts = TypeVarTuple("Ts")
Ts2 = TypeVarTuple("Ts2")

def concat(a: tuple[*Ts], b: tuple[*Ts2]) -> tuple[*Ts, *Ts2]:
    ...
```

Another language with variadic generics is C++ which does support this:
```cpp
template<typename... A, typename... B>
auto concat(std::tuple<A...>, std::tuple<B...>) -> std::tuple<A..., B...>;
```

**Comment by erictraut** [comment]
The problem is that if a type such as `tuple[*Ts, *Ts2]` were used to annotate an input parameter, there would be no way to determine which arguments map to which `TypeVarTuple`. 

```python
def func(*a: *tuple[*Ts, *Ts2]) -> Union[*Ts, *Ts2]:
    ...

func(1, 2, 3, 4)
```

**Reply by Zomatree** [reply]
C++ requires that one of them be explicit or deduced:
![Code_-_Insiders_2BdIeLFdKv](https://github.com/python/typing/assets/39768508/6c23142f-aa4a-4171-b0ff-ff0a0defa57f)

```py
func[int, int](1, 2, 3, 4, 5)  # tuple[int, int] | tuple[int, int, int]
```

**Comment by Viicos** [comment]
> The problem is that if a type such as `tuple[*Ts, *Ts2]` were used to annotate an input parameter, there would be no way to determine which arguments map to which `TypeVarTuple`.

This is a valid when using `TypeVarTuple` indeed. However, I'm wondering if it could at least be supported when not using such variables. My use case is I'm trying to type a function taking `*args`, where at least one of the arguments must be of a certain type, in any order (let's say `int`):

```python
# Matches `func(1, "a", "b", ...)
@overload
def func(arg: int, *args: str): ...

# Matches `func("a", "b", ..., 1)
@overload
def func(*args: *tuple[*tuple[str, ...], int]): ...

# Matches `func("a", "b", ..., 1, "a", "b", ...), but not valid currently
@overload
def func(*args: *tuple[*tuple[str, ...], int, *tuple[str, ...]]): ...
```
